### PR TITLE
[resource/host] Add semantic convention for IP addresses of a host

### PR DESCRIPTION
**Pull request by mx-psi**
~~Adds `host.ipv4.addresses` and `host.ipv6.addresses` resource semantic conventions to specify the IPv4 and IPv6 addresses of a host.~~
 
Adds `host.ip` resource attribute for specifying the IPs of a host, ensuring well-recognized formats for IPv4 and IPv6 addresses

Updates #131 (missing `host.mac`)

Reference implementation on `system` detector: open-telemetry/opentelemetry-collector-contrib/pull/24450

### CODE REVIEW

File: docs/resource/host.md [hunk 1]
```diff
@@ -16,6 +16,12 @@
 | `host.image.name` | string | Name of the VM image or OS install the host was instantiated from. | `infra-ami-eks-worker-node-7d4ec78312`; `CentOS-8-x86_64-1905` | Recommended |
 | `host.image.id` | string | VM image ID or host OS image ID. For Cloud, this value is from the provider. | `ami-07b06b442921831e5` | Recommended |
 | `host.image.version` | string | The version string of the VM image or host OS as defined in [Version Attributes](README.md#version-attributes). | `0.1` | Recommended |
+| `host.ipv4.addresses` | string[] | Available IPv4 addresses of the host, excluding loopback interfaces. [1] | `[['192.168.1.140']]` | Opt-In |
+| `host.ipv6.addresses` | string[] | Available IPv6 addresses of the host, excluding loopback interfaces. [2] | `[['fe80::abc2:4a28:737a:609e', 'fe80::849e:eaff:fe31:3c90']]` | Opt-In |
+
+**[1]:** Addresses MUST be specified in dotted-quad notation.
```
Comment (mx-psi) [inline docs/resource/host.md:22]: Quoting [Wikipedia](https://en.wikipedia.org/wiki/Dot-decimal_notation),

> No formal specification of this textual IP address representation exists

The Wikipedia article also links [this expired draft](https://datatracker.ietf.org/doc/html/draft-main-ipaddr-text-rep-02).

**Comment by mx-psi** [comment]
> Do we need the explicit differentiation of `ipv4` and `ipv6` addresses through different attributes? What would be the use cases to explicitly differentiating those?

We discussed this in the system semantic convention WG as well as in https://github.com/open-telemetry/semantic-conventions/issues/131#issuecomment-1643439369. Personally the main benefit I see is that we make semantics of the attributes more specific; I have been bit by very general attributes like `host.name` and `host.id` more than once in the past and want to avoid that generally. What is the benefit of being ambiguous?

Discussed https://github.com/open-telemetry/semantic-conventions/issues/131#issuecomment-1643439369
> Also, in [ECS we have the `host.ip`](https://www.elastic.co/guide/en/ecs/current/ecs-host.html) (array-typed) field. Couldn't we just go with that?

I find it at least possible that `host.ip` (or `host.ipv4`/`host.ipv6`) would be used as a namespace in the future and thus I think we should avoid just `host.ip` per the "Names SHOULD NOT coincide with namespaces" guidance [here](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.22.0/specification/common/attribute-naming.md)

**Comment by Oberon00** [comment]
The v4/v6 name split is also inconsistent with our network.* semantic  conventions.

**Comment by AlexanderWert** [comment]
> We discussed this in the system semantic convention WG as well as in [#131 (comment)](https://github.com/open-telemetry/semantic-conventions/issues/131#issuecomment-1643439369). Personally the main benefit I see is that we make semantics of the attributes more specific; I have been bit by very general attributes like `host.name` and `host.id` more than once in the past and want to avoid that generally. What is the benefit of being ambiguous?

One benefit of being "ambiguous" here is that querying and interoperating with the data is easier, i.e. does not require knowledge about the IP version. For example, let's say you want to correlate data by an IP.  With the above proposal, before querying data you'd need to parse the value of an IP first to understand which version it is, before being able to query on the right attribute (`host.ipv4.addresses` vs `host.ipv6.addresses`). 

Attributes with comparable, *intentional* ambiguity are [`server.address`](https://opentelemetry.io/docs/specs/semconv/general/general-attributes/#server-attributes), `client.address`, `source.address`, `destination.address`, etc. For the above reasons,  in those cases the attributes are intentionally defined in a non-prescriptive way so that `server.address` can be either a FQDN or an IP, or the `destination.address` an IP or a UNIX socket name.

> I find it at least possible that host.ip (or host.ipv4/host.ipv6) would be used as a namespace in the future and thus I think we should avoid just host.ip per the "Names SHOULD NOT coincide with namespaces" guidance [here](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.22.0/specification/common/attribute-naming.md)

Interesting, haven't thought about that. Do you have examples of additional attributes that we'd like to have under an `host.ip` namespace?

**Comment by Oberon00** [comment]
> The v4/v6 name split is also inconsistent with our network.* semantic conventions.
>
>> Would you mind elaborating on which conventions are inconsistent with these and how?

@mx-psi See here: https://github.com/open-telemetry/semantic-conventions/blob/main/docs/general/attributes.md#network-attributes

Especially `source.address`,  `destination.address`, `server.address`, `client.address`, `client.socket.address`, `server.socket.address`.

**Comment by mx-psi** [comment]
> One benefit of being "ambiguous" here is that querying and interoperating with the data is easier, i.e. does not require knowledge about the IP version. For example, let's say you want to correlate data by an IP. With the above proposal, before querying data you'd need to parse the value of an IP first to understand which version it is, before being able to query on the right attribute (host.ipv4.addresses vs host.ipv6.addresses).

@AlexanderWert (also @Oberon00) My main goal was to ensure that if someone wants to determine whether an address is IPv4/IPv6 they can unambiguously recover that information.

I think Alex's is a fair point and prior attributes are also a good argument against splitting per version. A compromise would be to have a single attribute but provide strict guidance on what the IP format should be for each version (current wording includes references to "quad-dotted notation" and to RFC 5952, that seems enough to me) so that one can still recover the IP version. Would that be okay with you?

> Interesting, haven't thought about that. Do you have examples of additional attributes that we'd like to have under an host.ip namespace?

I don't have anything particular in mind (maybe someday we want to express CIDR blocks? or something about IP headers?), but I wouldn't want my lack of imagination to put us in the future in a spot where we have to contradict the spec guidance when we can easily avoid it.

File: model/resource/host.yaml [hunk 2]
```diff
@@ -70,3 +70,19 @@ groups:
...
+          Available IPv4 addresses of the host, excluding loopback interfaces.
+        note: >
+          Addresses MUST be specified in dotted-quad notation.
+        examples: ["192.168.1.140"]
+      - id: ipv6.addresses
+        type: string[]
+        requirement_level: opt_in
+        brief: >
+          Available IPv6 addresses of the host, excluding loopback interfaces.
+        note: >
```
Comment (dineshg13) [inline model/resource/host.yaml:86]: can you please add an example here as well ?

**Comment by ChrsMark** [comment]
Hey @mx-psi, I'm +1 for the one single attribute (`host.ip`). Regarding the potential additional attributes I'm not sure how possible that would be so I would avoid converting the convenient/compact `host.ip` to `host.ip.addresses`. I would just stick with the ECS version of the field :).

**Comment by mx-psi** [comment]
I merged both attributes into `host.ip`. I still think `host.ip.addresses` would be slightly better in terms of future-proofing but I understand conciseness is also a factor here. I left the wording regarding formatting of IPv4 and IPv6 addresses.

Do we have any guidance regarding plurals in attribute names? For example, we have `process.command_args` and in general we use plurals elsewhere. Should we do that here (so that would be `host.ips`)?

**Comment by MikePaquette** [comment]
Hi Folks,
Mike P. from Elastic jumping in here with a few thoughts:

> I merged both attributes into host.ip.


+1  In general, it’s best for existing users of ECS if we maintain compatibility (field name equivalence) with widely deployed ECS fields, like `host.ip` unless there is a strong reason to do otherwise.

> I still think host.ip.addresses would be slightly better in terms of future-proofing but I understand conciseness is also a factor here. I left the wording regarding formatting of IPv4 and IPv6 addresses.


I hope we can avoid all situations where we  might add an extension (like `.addresses`) to an existing ECS field.  This provides two benefits to our users 1) a consistent practice where a “leaf” field name is always a leaf field name.  (e.g., `host.ip`, `source.ip`, `client.ip`, `destination.ip`, etc.). - this helps analysts remember and/or deduce field names.  2) It avoids potential field data type mismatches in Elasticsearch where a given field name cannot be an [object](https://www.elastic.co/guide/en/elasticsearch/reference/current/object.html) and a “leaf” field in the same index.  During the development of ECS, we made one such error, making `host` an object (i.e., the leading part of `host.ip`), when previously Logstash used `host` as a “leaf” field name.  We had to deal with reconciling these mismatches for years in mixed user environments.

> Do we have any guidance regarding plurals in attribute names? For example, we have process.command_args and in general we use plurals elsewhere. Should we do that here (so that would be host.ips)?

ECS provides a set of [guidelines for field names](https://www.elastic.co/guide/en/ecs/8.9/ecs-guidelines.html#_guidelines_for_field_names), which includes the rather obvious “Use singular and plural names properly to reflect the field content. For example, use `requests_per_sec` rather than `request_per_sec`.”, but this is not very helpful in this case, as we all know that a host can indeed have multiple IP addresses.  ECS has another guideline which is to avoid abbreviations where possible, except when the name used for the concept is **widely accepted in the abbreviation form**, such as `ip` field, or field sets like `os` or `geo`.  In this case, given the widespread acceptance of `ip`, and the ambiguity of `ips` (Intrusion Prevention System, anyone?) the familiar `ip` would win out.

If we anticipate that some systems adopting Otel SemConv will not support a combined IPv4/IPv6 field data type, then we could entertain proposals to add new fields, such as `host.ipv4` and `host.ipv6`, but `host.ip` should always continue to be populated.

**Review by ChrsMark — APPROVED** [review]
LGTM. I have filed https://github.com/open-telemetry/opentelemetry-specification/pull/3641 to further discuss and clarify the pluralization guidelines.

File: model/resource/host.yaml [hunk 3]
```diff
@@ -70,3 +70,12 @@ groups:
           The version string of the VM image or host OS as defined in
           [Version Attributes](README.md#version-attributes).
         examples: ['0.1']
+      - id: ip
```
Comment (lmolkova) [inline model/resource/host.yaml:73]: we avoided using `IP` in the server/client namespaces and used `address`  as it allows covering other kinds of socket addresses. I'd propose to use `host.socket.address` here as well.

Comment (Oberon00) [reply model/resource/host.yaml:73]: In the context of a particular operation, only one kind of address will be relevant. But a host will have multiple address kinds which are not all distinguishable by syntax. So I think ip is fine in the name.

Comment (AlexanderWert) [reply model/resource/host.yaml:73]: I agree with @Oberon00 's comment above.

Plus, I assume in the overwhelming majority of cases the addresses users care about would be IP. So, from a user perspective `host.socket.address` would be not the most intuitive attribute to search for when someone would want to know the IP addresses of a host.

Instead, I'd propose to add additional fields for other families of socket addresses, if we want to model them on the host resource. But let's do it based on practical examples / use cases and a based on a concrete need.

**Comment by joaopgrassi** [comment]
I generally also agree that having attributes separated by version would hurt how one search for it - you would need to either know which version is being used OR do a query with both and see which one pop up with data so having a single seems better to me. 

What I'm not so sure is using a singular term for an array, since that goes against our recommendation. Adding an exception just opens up for more interpretations to "what is wildly known" (https://github.com/open-telemetry/opentelemetry-specification/pull/3641/files). 

In another hand, we said we would take ECS fields and using something other than `host.ip` will break it. I guess either way we will have to make a compromise - either compromise OTel naming conventions with plural attributes and go with `host.ip` or compromise ECS and go with `host.ip.addresses` or `host.ip_addresses` or the like.

**Comment by mx-psi** [comment]
If we compromise ECS, I would favor `host.ip.addresses` instead of `host.ip_addresses`, since it seems at least conceivable to me that we may want to add further items to a `host.ip` namespace at some point in the future.

**Comment by MikePaquette** [comment]
Sorry for the delay, catching up from some time off

> What I'm not so sure is using a singular term for an array, since that goes against our recommendation. Adding an exception just opens up for more interpretations to "what is wildly known" (https://github.com/open-telemetry/opentelemetry-specification/pull/3641/files).

If our focus is on the success of analysts interacting with our data, which I think it is, then we should favor exceptions for widely-known terms over strict-adherence to rules like **_all array type field names must use plural form_**.  

While strict adherence can have some advantages when it comes to automated validation of fields in a schema, in this case, since English language supports irregular plurals, even a simple automated check such as “if type = array, check name ends in ’s’” will require an exception path for irregular plurals, for example, child-children, mouse-mice, radius-radii, etc., so if we’d need to have exceptions for automated checks anyway, it makes more sense to apply the exception in favor of the consumer of our data. 

> In another hand, we said we would take ECS fields and using something other than host.ip will break it. I guess either way we will have to make a compromise - either compromise OTel naming conventions with plural attributes and go with host.ip or compromise ECS and go with host.ip.addresses or host.ip_addresses or the like. I guess we need to pick the one that causes less problems for both sides.

From the Elastic side, we are strongly in favor of making the exception to the plural attributes naming convention, and avoiding the breaking change.  We think that current and future users will appreciate the consistency.

> Member
> Author
> [mx-psi](https://github.com/mx-psi) commented [2 weeks ago](https://github.com/open-telemetry/semantic-conventions/pull/203#issuecomment-1677072126)
> If we compromise ECS, I would favor host.ip.addresses instead of host.ip_addresses, since it seems at least conceivable to me that we may want to add further items to a host.ip namespace at some point in the future.

If we can’t agree with the above recommendation to make the exception to the plural attributes naming convention, and we feel we must make a breaking change for the thousands of users already using ECS `host.ip`, I would again ask us to consider that there are two kinds of breaking changes:
- A name change only, whereby a new field, with the same semantics is created - e.g., `host.ip_addresses` - this could be represented as an “alias” for `host.ip` in environments as they migrate from ECS to the merged schema.
- A change that creates potential mapping conflicts between Elasticsearch “leaf” and “object” fields - e.g., `host.ip.addresses` - We have learned from experience that this type of change is particularly troublesome, and difficult to work around, so we would ask strongly to never “extend” existing leaf fields like `.ip` by adding a `.anything`.

**Comment by mx-psi** [comment]
I want to discuss this on next week's Specification SIG meeting, the main point of contention seems to be if/how much we should deviate from ECS, and I think a general framework like what is discussed on https://github.com/open-telemetry/semantic-conventions/pull/181#discussion_r1312729412 would help move this forward. Please attend if you want/can so we can move this foward :)

**Comment by Oberon00** [comment]
I don't have a horse in the race, but based on the pros/cons, I it seems logical to suggest:

### D) `host.inet.addresses`

++

1. We honor the plural vs singular rule
2. We reserve inet as a namespace for ip-related attributes as a namespace for the future


--

1. We break ECS users but only on naming not the schema
2. We use a slightly unusual namespace (inet is used for IP in POSIX/libc (https://www.gnu.org/software/libc/manual/html_node/Internet-Address-Formats.html) other alternatives not equal to ip would also work)


Having said that, I agree that host.ip seems fine, and host.inet or the other alternative can then be used later on if we do find we need more ip-related attributes other than the ip address itself.

**Review by joaopgrassi — APPROVED** [review]

**Comment by joaopgrassi** [comment]
@open-telemetry/specs-semconv-approvers @open-telemetry/specs-semconv-maintainers We need more reviews on this, given the discussions about ECS and etc.

**Review by pyohannes — APPROVED** [review]

**Review by frzifus — COMMENTED** [review]
Damn I guess I am a bit late.

> > I still think host.ip.addresses would be slightly better in terms of future-proofing but I understand conciseness is also a factor here. I left the wording regarding formatting of IPv4 and IPv6 addresses.
> 
> I hope we can avoid all situations where we might add an extension (like `.addresses`) to an existing ECS field. This provides two benefits to our users 1) a consistent practice where a “leaf” field name is always a leaf field name. (e.g., `host.ip`, `source.ip`, `client.ip`, `destination.ip`, etc.). - this helps analysts remember and/or deduce field names. 2) It avoids potential field data type mismatches in Elasticsearch where a given field name cannot be an [object](https://www.elastic.co/guide/en/elasticsearch/reference/current/object.html) and a “leaf” field in the same index. During the development of ECS, we made one such error, making `host` an object (i.e., the leading part of `host.ip`), when previously Logstash used `host` as a “leaf” field name. We had to deal with reconciling these mismatches for years in mixed user environments.
> 

While I see this point, I am a bit confused. Isnt the current spec promoting fields like `client.address` or `server.address`  - I noticed this from the http space https://github.com/open-telemetry/opentelemetry-specification/issues/3163? In that case its clear that we expect a single address. And here / in ECS e.g. [destination](https://www.elastic.co/guide/en/ecs/current/ecs-destination.html) and [server](https://www.elastic.co/guide/en/ecs/current/ecs-server.html#field-client-ip) provide an `address` field. I assume that a spec consistent with its recommendations would also help e.g. analyst too.

**Pull request merged**